internal/controller/controllers/agent_controller.go

```diff
@@ -80,69 +97,244 @@ func (r *AgentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	// check for updates from user, compare spec and update if needed
-	result, err = r.updateIfNeeded(ctx, agent, cluster)
+	err = r.updateIfNeeded(ctx, agent, cluster)
 	if err != nil {
-		r.updateFailure(ctx, agent, err)
-		return result, nil
+		return r.updateStatus(ctx, agent, host, err, !IsHTTP4XXError(err))
 	}
 
-	err = r.updateInventory(cluster, agent)
+	err = r.updateInventory(host, agent)
 	if err != nil {
-		r.updateFailure(ctx, agent, err)
-		return ctrl.Result{Requeue: true}, nil
+		return r.updateStatus(ctx, agent, host, err, true)
+	}
+
+	return r.updateStatus(ctx, agent, host, nil, false)
+}
+
+func (r *AgentReconciler) updateStatus(ctx context.Context, agent *adiiov1alpha1.Agent, h *models.Host, syncErr error, internal bool) (ctrl.Result, error) {
+
+	r.specSynced(agent, syncErr, internal)
+
+	if h != nil && h.Status != nil {
+		status := *h.Status
+		r.connected(agent, status)
+		r.readyForInstallation(agent, status)
+		r.validated(agent, status, h)
+		r.installed(agent, status, swag.StringValue(h.StatusInfo))
+	} else {
+		r.setConditionsUnknown(agent)
 	}
+	if updateErr := r.Status().Update(ctx, agent); updateErr != nil {
+		r.Log.WithError(updateErr).Error("failed to update agent Status")
+		return ctrl.Result{Requeue: true}, updateErr
+	}
+	if internal {
+		return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, nil
+	}
+	return ctrl.Result{}, nil
+}
 
+func (r *AgentReconciler) setConditionsUnknown(agent *adiiov1alpha1.Agent) {
 	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
-		Type:    adiiov1alpha1.AgentSyncedCondition,
-		Status:  corev1.ConditionTrue,
-		Reason:  adiiov1alpha1.AgentSyncedReason,
-		Message: adiiov1alpha1.AgentStateSynced,
+		Type:    adiiov1alpha1.InstalledCondition,
+		Status:  corev1.ConditionUnknown,
+		Reason:  adiiov1alpha1.NotAvailableReason,
+		Message: adiiov1alpha1.NotAvailableMsg,
 	})
-	if updateErr := r.Status().Update(ctx, agent); updateErr != nil {
-		r.Log.WithError(updateErr).Error("failed to update agent status")
-		return ctrl.Result{Requeue: true}, nil
+	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
+		Type:    adiiov1alpha1.ConnectedCondition,
+		Status:  corev1.ConditionUnknown,
+		Reason:  adiiov1alpha1.NotAvailableReason,
+		Message: adiiov1alpha1.NotAvailableMsg,
+	})
+	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
+		Type:    adiiov1alpha1.ReadyForInstallationCondition,
+		Status:  corev1.ConditionUnknown,
+		Reason:  adiiov1alpha1.NotAvailableReason,
+		Message: adiiov1alpha1.NotAvailableMsg,
+	})
+	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
+		Type:    adiiov1alpha1.ValidatedCondition,
+		Status:  corev1.ConditionUnknown,
+		Reason:  adiiov1alpha1.NotAvailableReason,
+		Message: adiiov1alpha1.NotAvailableMsg,
+	})
+}
+
+func (r *AgentReconciler) specSynced(agent *adiiov1alpha1.Agent, syncErr error, internal bool) {
```

Comment:
Here also it would help to explain how `internal` is meant to be used.

Reply:
Done